spinalcordtoolbox/scripts/sct_maths.py

```diff
@@ -21,71 +23,93 @@
-            try:
-                float(element)
-                return True
-            except ValueError:
-                return False
-
-        # '-i' is mandatory, and the first argument in the parser, so this *should* be parsed first
-        data_in = Image(namespace.i).data
-
-        data_out = []
-        for val in values:
-            # Case 1: Argument is a float
-            if is_float(val):
-                data_out.append(np.full_like(data_in, float(val)))
-            # Case 2: If argument is not a float, assume argument is a path to an image file
-            else:
-                data = Image(val).data
-                if data.shape == data_in.shape:
-                    data_out.append(data)
-                else:
-                    parser.error(f"Dimensions of '{val}' ({data.shape}) "
-                                 f"must match input image ({data_in.shape}).")
-
-        # Make sure the custom action follows the same logic as the regular "store" action
-        store(namespace, dest=self.dest, values=data_out)
-
-
-def store(namespace, dest, values):
+    try:
+        return float(arg)
+    except ValueError:
+        return arg
+
+
+def denoise_params(arg: str) -> tuple[int, int]:
+    """
+    Parse the argument for `-denoise` into a pair (patch radius, block radius).
+
+    Any sub-arguments that don't start with "p=" or "b=" are silently ignored.
     """
-    In addition to the default behavior (storing the argument value in the `arguments` namespace),
-    we preserve the parsed arguments in an ordered list. This allows us to execute each of the `sct_maths`
-    operations in the order they were passed in, while also allowing for duplicate calls to the same argument.
+    p, b = 1, 5  # defaults
+    for sub_arg in arg.split(","):
+        if sub_arg.startswith("p="):
+            p = int(sub_arg[2:])
+        elif sub_arg.startswith("b="):
+            b = int(sub_arg[2:])
+    return (p, b)
+
+
+class StoreTodo(argparse.Action):
+    """
+    Store the arguments of an sct_maths operation in `arguments.todo`.
+
+    The format is: (operation name, positional args, keyword args).
     """
-    setattr(namespace, dest, values)    # default behavior of argparse._StoreAction.__call__
-    setattr(namespace, "ordered_args",  # store the parsed argument in a separate list of ordered args
-            getattr(namespace, "ordered_args", []) + [(dest, values)])
+    def ___call___(self, parser, namespace, values, option_string=None):
+        # make sure values is a list, rather than a single bare value
+        if self.nargs in [None, '?']:
+            values = [values]
+
+        namespace.todo.append((self.dest, values, {}))
+
+
+class StoreTodoWithShapeDim(argparse.Action):
+    """
+    Store the arguments of `-dilate` or `-erode` in `arguments.todo`.
+
+    These two operations are special-cased because they need the "current"
+    values of `-shape` and `-dim` to also be saved.
```

Comment:
I also like your previous suggestion of putting `shape` and `dim` information directly into the argument for `-dilate` and `-erode`, but that's not backwards compatible, so I thought it was maybe out of scope.